### PR TITLE
Make the allocators' unittests and allocate, owns and resolveInternal…

### PR DESCRIPTION
**Pull request by edi33416**
…Pointer methods safe

Allocating memory should be a safe operation and so should functions like owns and resolveInternalPointer.

### CODE REVIEW

File: std/experimental/allocator/building_blocks/affix_allocator.d [hunk 1]
```diff
@@ -88,7 +88,7 @@ struct AffixAllocator(Allocator, Prefix, Suffix = void)
     private template Impl()
     {
 
-        size_t goodAllocSize(size_t s)
+        @safe size_t goodAllocSize(size_t s)
```
Comment (JackStouffer) [inline std/experimental/allocator/building_blocks/affix_allocator.d]: There's no guarantee that `parent.goodAllocSize` is `@safe`. Best leave this to be inferred.

File: std/experimental/allocator/building_blocks/affix_allocator.d [hunk 2]
```diff
@@ -168,23 +172,26 @@ struct AffixAllocator(Allocator, Prefix, Suffix = void)
         }
 
         static if (hasMember!(Allocator, "owns"))
-        Ternary owns(void[] b)
+        @safe Ternary owns(void[] b)
```
Comment (JackStouffer) [inline std/experimental/allocator/building_blocks/affix_allocator.d]: same problem here

**Comment by JackStouffer** [comment]
Under no circumstances should `@trusted` be applied to an entire function.

**Comment by aG0aep6G** [comment]
I've protested the change to `ubyte[]` with code that looks innocuous but actually causes memory corruption (see [here](https://github.com/dlang/phobos/pull/5322#issuecomment-291524917)). You're now making that code `@safe`.

```d
import core.memory: GC;
import std.stdio: writeln;
import std.experimental.allocator: make;
import std.experimental.allocator.building_blocks: NullAllocator, Region;

struct S { int* p; }

void main() @safe /* !! */
{
    auto buf = new ubyte[](S.sizeof);
    auto r = Region!(NullAllocator, 1)(buf);
    S* s = r.make!S(new int);
    *(*s).p = 42;
    writeln(*(*s).p); /* prints "42" as expected */
    
    version (none) GC.collect(); /* not @safe */
    else new int[](1_000); /* trigger a GC collection in an @safe manner */
    
    *(new int) = 13;
    writeln(*(*s).p); /* prints "13" - whoopsie */
}
```

This is exactly what must not happen in `@safe` code.

File: std/experimental/allocator/building_blocks/affix_allocator.d (hunk 1 above)
Comment (edi33416) [inline std/experimental/allocator/building_blocks/affix_allocator.d]: Imho I think that this function should be safe and thus I thought we should enforce it upon the parent.

File: std/experimental/allocator/building_blocks/affix_allocator.d (hunk 2 above)
Comment (edi33416) [inline std/experimental/allocator/building_blocks/affix_allocator.d]: same for this

**Comment by andralex** [comment]
OK @edi33416 and I just talked about this and we'll go with the following idea - define this struct:

```d
struct OpaquePointer
{
    private void* pointer;
    @disable this(void*);
}
```

Now, `new OpaquePointer[100]` will mark the memory for scanning by the GC.

Allocators that manage a raw buffer of memory will accept `OpaquePointer[]` safely. The responsbility of allocating the buffer properly falls on the caller.

@edi33416 you may want to shorten this PR to only include the other, simpler primitives.

**Comment by andralex** [comment]
this is being sold as parts :)

**Pull request closed without merging**